libs/sdk-ui-charts/src/highcharts/chartTypes/sankeyChart/sankeyChartOptions.ts

```diff
@@ -0,0 +1,83 @@
+// (C) 2023 GoodData Corporation
+import { ISeriesItem } from "../../typings/unsafe";
+import { DataViewFacade, getMappingHeaderFormattedName, IColorAssignment } from "@gooddata/sdk-ui";
+import { IUnwrappedAttributeHeadersWithItems } from "../../typings/mess";
+import { getColorByGuid, getRgbStringFromRGB, valueWithEmptyHandling } from "@gooddata/sdk-ui-vis-commons";
+import { IColorPalette, isRgbColor } from "@gooddata/sdk-model";
+import compact from "lodash/compact";
+
+const KEYS = ["from", "to", "weight", "name"];
+const DEFAULT_ATTRIBUTE_HEADER = "";
+const DEFAULT_WEIGHT = 1;
+
+const buildNodes = (
+    colorAssignments: IColorAssignment[],
+    colorPalette: IColorPalette,
+    emptyHeaderTitle: string,
+) => {
+    return colorAssignments.map((colorAssignment) => ({
+        id: valueWithEmptyHandling(
+            getMappingHeaderFormattedName(colorAssignment.headerItem),
+            emptyHeaderTitle,
+        ),
+        color: isRgbColor(colorAssignment.color)
+            ? getRgbStringFromRGB(colorAssignment.color.value)
+            : getRgbStringFromRGB(getColorByGuid(colorPalette, colorAssignment.color.value, 0)),
+    }));
+};
+
+const buildEmptyData = (dv: DataViewFacade) => {
+    return [
+        {
+            from: DEFAULT_ATTRIBUTE_HEADER,
+            to: DEFAULT_ATTRIBUTE_HEADER,
+            weight: Number((dv.rawData().dataAt(0) as Array<number>)[0]) || DEFAULT_WEIGHT,
+            name: DEFAULT_ATTRIBUTE_HEADER,
+        },
+    ];
+};
+
+const buildData = (
+    dv: DataViewFacade,
+    attributeHeaders: IUnwrappedAttributeHeadersWithItems[],
+    emptyHeaderTitle: string,
+) => {
+    const [from, to] = compact(attributeHeaders);
+    const isExistingAttributeHeader = from || to;
+
+    if (!isExistingAttributeHeader) {
+        return buildEmptyData(dv);
+    }
+
+    return (dv.rawData().dataAt(0) as Array<number>).map((it, index) => {
+        const fromValue =
+            from &&
+            valueWithEmptyHandling(getMappingHeaderFormattedName(from.items[index]), emptyHeaderTitle);
```

Comment:
I am currently sending dimension data as follows:

```typescript
newTwoDimensional(
    [MeasureGroupIdentifier],
    compact([...attributeFromByAttributes, ...attributeToByAttributes]),
);
```
With the dimensions sent as above, the SQL created to retrieve data will be:

```SQL
select
    attribute_from, attribute_to, metric
where ...
group by 
    attribute_from, attribute_to
```
I have checked the generated query on Splunk and it is as expected. With this query, the data will return with the same number of items for attribute_from, attribute_to, and metric.

In fact, I have consulted the combo chart in case of sending 1 metric and 2 attributes in the rows bucket to reuse for the Sankey chart. As someone with little experience with AFM, could you share with me some cases where the data returned has different numbers of attr_from and attr_to?

Thank you very much.

Reply:
ah, I see now, thanks for the explanation